src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -1181,6 +1183,7 @@ def filter(self, col: Expression) -> "SelectStatement":
             new = SelectStatement(
                 from_=self.to_subqueryable(), where=col, analyzer=self.analyzer
             )
+        new._attributes = self._attributes
```

Comment:
Wouldn't this also apply to set operations like union and intersection?

Reply:
Can we first check if the input data types are the same, then it can be propagated to the output. I believe this would be the more common case, and we should be able to recognize it.

Reply:
Actually it's [data coerction](https://docs.snowflake.com/en/sql-reference/data-type-conversion#implicit-casting-coercion) in Snowflake, and there are many [complicated rules](https://docs.snowflake.com/en/sql-reference/data-type-conversion#data-types-that-can-be-cast) on the server side. Even if they are the same type, there may still be data coercion happening (such as two varchar values with different length). So I don't think it's a good idea to have such rules on the client side, and we should only rely on the server side for data coercion. This type of operation is not in the scope of reducing describe queries.

Reply:
Hmm, I think this will reduce describe queries. We're just worried that the rules we use on the client side may end up being different than the server side, right? Would we still be worried if the data types for both operands are exactly the same (including lengths for archer, etc)? Or is it that we don't have this information on the client side so we cannot be 100% sure? I'm just trying to understand the concern.

Reply:
My original concern is just about data coercion, but as you said, if two types are exactly the same, we can try to infer too and we can investigate it in the future, if this is a common pattern that we should optimize.

Reply:
Yes, I agree it doesn't have to be in this PR. For Snowpark pandas, we do use union operations pretty commonly. And I think that in most cases, the data types are identical.

Reply:
got it, yeah if it's pretty common, we can definitely do it.